File: BI-TJV/semestral/rest-server/src/main/java/wrzecond/service/TjvService.java
package wrzecond.service;

import lombok.AllArgsConstructor;
import org.springframework.data.domain.Sort;
import wrzecond.ICreateDTO;
import wrzecond.IReadDTO;
import wrzecond.entity.TjvEntity;
import wrzecond.repository.TjvRepository;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

@AllArgsConstructor
public abstract class TjvService<T extends TjvEntity, R extends IReadDTO, C extends ICreateDTO> {

    // === PARAMETERS ===
    private final TjvRepository<T> repository;

    // === GET METHODS ===
    protected T getEntityByID (Integer id) {
        return repository.getOne(id);
    }
    public R getByID (Integer id) {
        return toDTO(getEntityByID(id));
    }
    public List<T> findByIDs (List<Integer> ids) {
        return ids.stream()
                .map(this::getEntityByID)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }
    public List<R> findAll () {
        return repository.findAll(getSort())
                .stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // === POST METHODS ===
    public R create (C dto) {
        T entity = repository.save(fromDTO(dto));
        repository.flush();
        return toDTO(entity);
    }
    public List<R> create (List<C> dtoList) {
        List<T> entities = dtoList.stream()
            .map(this::fromDTO)
            .map(repository::save)
            .collect(Collectors.toList());
        repository.flush();
        return entities.stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public void update (Integer id, C dto) {
        T entity = repository.getOne(id);
        repository.save(merge(entity, dto));
        repository.flush();
    }

    // === DELETE METHODS ===
    public void delete (Integer id) {
        repository.deleteById(id);
    }

    // === SORT ===
    protected Sort getSort () {
        return Sort.unsorted();
    }

    // === CONVERSION ===
    protected abstract R toDTO (T entity);
    protected abstract T fromDTO (C dto);
    protected abstract T merge (T entity, C dto);

}

File: BI-KOT/semestral/rest-server/src/main/kotlin/wrzecond/service/TjvService.kt
package wrzecond.service

import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.orm.jpa.JpaObjectRetrievalFailureException
import org.springframework.web.server.ResponseStatusException
import wrzecond.ICreateDTO
import wrzecond.IReadDTO
import wrzecond.IUpdateDTO
import wrzecond.entity.TjvEntity
import wrzecond.repository.TjvRepository

abstract class TjvService<T: TjvEntity, R: IReadDTO, C: ICreateDTO, U: IUpdateDTO> (open val repository: TjvRepository<T>) {

    /** Default sort */
    protected open val sort: Sort
        get () = Sort.unsorted()

    /** Helper method to get entity with given ID */
    fun getEntityByID  (id: Int) = tryCatch { repository.getOne(id) }

    // Get
    fun getByID (id: Int) = getEntityByID(id).toDTO()
    fun findByIDs (ids: List<Int>) = ids.map(this::getEntityByID)
    fun findAll () = repository.findAll(sort).map { it.toDTO() }

    // Create, Update, Delete
    fun create (vararg dtos: C)         = tryCatch { dtos.map { repository.save(it.toEntity()).toDTO() }.apply { repository.flush() } }
    fun update (id: Int, dto: U) : Unit = tryCatch { repository.saveAndFlush(getEntityByID(id).merge(dto)) }
    open fun delete (id: Int)           = tryCatch { repository.deleteById(id) }

    // Extension function allowing for mapping Entity to Read DTO
    abstract fun T.toDTO() : R

    // Extension function allowing for mapping Create DTO to Entity
    abstract fun C.toEntity() : T

    // Extension function allowing for updating Entity with Update DTO
    abstract fun T.merge(dto: U) : T

    /**
     * Helper extension function allowing to perform repository action,
     * catch any exception and change it to ResponseStatusException
     * @param block action to perform
     * @return block return value
     */
    protected fun <X> tryCatch (block: TjvRepository<T>.() -> X) =
        try { repository.block() }
        catch (_: JpaObjectRetrievalFailureException) { throw ResponseStatusException(HttpStatus.NOT_FOUND) }
        catch (_: EmptyResultDataAccessException)     { throw ResponseStatusException(HttpStatus.NOT_FOUND) }
        catch (e: Exception) { e.printStackTrace(); throw ResponseStatusException(HttpStatus.BAD_REQUEST) }

}
